server/app/middlewares/websocket_backend.rb

```diff
@@ -183,30 +184,29 @@ def find_node(req)
   # @param [Faye::WebSocket] ws
   # @param [Rack::Request] req
   def on_open(ws, req)
+    node_id = req.env['HTTP_KONTENA_NODE_ID']
+    connected_at = nil
+
     node = find_node(req)
 
+    send_master_info(ws)
```

Comment:
Always sending the `/agent/master_info` this early has a slight effect on the agent behavior in terms of when the `websocket:connected` notification is emitted. Currently the server might close the websocket with a 4010 invalid version websocket close immediately after the master_info, but the agent will shut down in this case. Now the agent will also get the master_info immediately before the 4020 invalid clock websocket close, and the agent will not shutdown, but reconnect instead.

AFAIK this just means that the agent log worker would bounce a lot if the agent has an invalid clock.

Reply:
> Is this because agent still gets master info before the connection reject and something starts to send Rpc messages already?

Yes, it will wake up the `LogWorker` and drain the queue, sending async RPC notifications, all of which now fail with a warning. In that specific case of invalid clocks, it used to get the close without the `master_info`.

I guess the best answer is to only explicit send the `master_info` where appropriate: the "invalid version" close (where it also breaks the logging, but it's necessary for the agent auto-upgrade to work) and then for the succesful `plugin!` case. I just wanted to get rid of the code duplication by moving it out of the `NodePlugger`, where it suffers from a race condition in the `handle_invalid_agent_version` case  :/